src/hotspot/share/oops/constantPool.cpp

```diff
@@ -167,10 +167,22 @@ objArrayOop ConstantPool::resolved_references_or_null() const {
   if (_cache == NULL) {
     return NULL;
   } else {
-    return (objArrayOop)_cache->resolved_references();
+    return _cache->resolved_references();
   }
 }
 
+oop ConstantPool::resolved_references_at(int index) const {
+  oop result = resolved_references()->obj_at(index);
+  assert(oopDesc::is_oop_or_null(result), "Must be oop");
+  return result;
+}
+
+// Use a CAS for multithreaded access
+oop ConstantPool::set_resolved_references_at(int index, oop new_result, oop old_result) {
```

Comment:
This function should be called `set_resolved_reference_at` - singular - as it sets one reference. Thanks.

Reply:
We never overwrite anything. I.e. we only store if the index is null.
From what I have seen in the code, anything else would be a bug?

So I think this method is a bit to generic:
bool set_resolved_reference_at(int index, oop value)
Where we return false is that index already contains an oop?

Reply:
The reason I made it resolved_references plural is so you can grep on resolved_references and see all operations on this array.

We do not overwrite anything.  We may have multiple threads writing the same string to it.  There's an assert in string_at_put:
  assert(result == nullptr || result == str, "Only set once or to the same string.");

I need it to return the existing value for the code in resolve_constant_at_impl().

Reply:
Shouldn't this method check the validity of the index argument?

Reply:
> I need it to return the existing value for the code in resolve_constant_at_impl().

So it's a getter/setter, but if the argument must be nullptr, I think we should skip it.

Reply:
obj_at_put already checks the index.

Reply:
I agree with David's naming. 

For grepping, it's better to grep for the singular form as there are already functions that use the singular form. 

```
$ find src/hotspot -type f | xargs grep resolved_reference | grep -v references | wc
     41     223    4559
```

Here are some examples. If you grep for the plural form, you will miss them.

```
./share/oops/constantPool.cpp:  set_resolved_reference_length(
./share/oops/constantPool.hpp:    int                _resolved_reference_length;
./share/oops/cpCache.hpp:  void initialize_resolved_reference_index(int ref_index) {
```